harvester/harvest/alma.py

```diff
@@ -0,0 +1,192 @@
+            filepath_date = convert_to_utc(date_parser(match.group(1)))
+
+            # skip if date out of bounds for harvester from/until dates
+            if self.from_datetime_object and filepath_date < self.from_datetime_object:
+                continue
+            if self.until_datetime_object and filepath_date >= self.until_datetime_object:
+                continue
+
+            yield filepath
+
+    def _list_s3_xml_files(self) -> list[str]:
+        """Return a list of S3 URIs for extracted Alma XML files.
+
+        Example self.input_files = "s3://timdex-extract-dev-222053980223/alma/"
+        """
+        bucket, prefix = self.input_files.replace("s3://", "").split("/", 1)
+        s3_objects = S3Client.list_objects_uri_and_date(bucket, prefix)
+        return [
+            s3_object[0]
+            for s3_object in s3_objects
+            if s3_object[0].lower().endswith(".xml")
+        ]
+
+    def _list_local_xml_files(self) -> list[str]:
+        """Return a list of local filepaths of extracted Alma XML files."""
+        return glob.glob(f"{self.input_files.removesuffix('/')}/*.xml")
+
+    def parse_marc_records_from_files(self) -> Iterator[MARCRecord]:
+        """Yield parsed MARCRecords from list of filepaths."""
+        for filepath in self._list_xml_files():
+            with smart_open.open(filepath, "rb") as f:
+                context = etree.iterparse(f, events=("end",), tag="record")
+                for _event, elem in context:
+                    yield MARCRecord(elem)
+                    elem.clear()
+                    while elem.getprevious() is not None:
+                        del elem.getparent()[0]
+
+    def filter_geospatial_marc_records(
+        self, marc_records: Iterator[MARCRecord]
+    ) -> Iterator[MARCRecord]:
+        """Yield geospatial MARC records by filtering on defined criteria."""
+        for i, record in enumerate(marc_records):
+            if i % 10_000 == 0 and i > 0:  # pragma: nocover
+                message = f"{i} MARC records scanned for geospatial filtering"
+                logger.info(message)
+
+            # skip if leader doesn't have a/c/n/p
+            if record.leader[5] not in ("a", "c", "n", "p"):
+                continue
+
+            # skip if Genre/Form 655 does not contain "Maps."
+            if not any(
+                "Maps." in subfield.value
+                for form in record["655"]
+                for subfield in form.subfield("a")
+            ):
+                continue
+
+            # skip if call number prefix not in list
+            if not any(
+                subfield.value in ("MAP", "CDROM", "DVDROM")
+                for locations in record["949"]
+                for subfield in locations.subfield("k")
+            ):
+                continue
+
+            # skip if shelving location not in list
+            if not any(
+                subfield.value in ("MAPRM", "GIS")
+                for locations in record["985"]
```

Comment:
[Non-blocking]: I continue to be curious about the schema from MIT-owned sources and wonder how we can gain access to the mappings. From my understanding, code 985 is a custom field, right?

Reply:
It's a great question, and I think related to @ehanson8's comment about stakeholder review and the complexity of this logic.

My vote would be to move forward with this logic as-is for now, knowing it still needs some stakeholder review.  Once we settle on the particular MARC fields and qualities that will filter to our desired records -- and maybe that _is_ this -- then I'd propose we add a standalone markdown doc that outlines the filtering criteria.